samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs

```diff
@@ -404,6 +409,83 @@ private bool TryExtractJsonFromOpenApiPlanResult(string openApiSkillResponse, ou
         return false;
     }
 
+    /// <summary>
+    /// Try to optimize json from the planner response
+    /// based on token limit
+    /// </summary>
+    private string OptimizeOpenApiSkillJson(string jsonContent, int tokenLimit, Plan plan)
+    {
+        int jsonTokenLimit = (int)(tokenLimit * 0.8);
+
+        // Remove all new line characters + leading and trailing white space
+        jsonContent = Regex.Replace(jsonContent.Trim(), @"[\n\r]", string.Empty);
+        var document = JsonDocument.Parse(jsonContent);
+        string lastSkillInvoked = plan.Steps[^1].SkillName;
+
+        // Check if the last skill invoked was GitHubSkill and deserialize the JSON content accordingly
+        if (string.Equals(lastSkillInvoked, "GitHubSkill", StringComparison.Ordinal))
```

Comment:
I think our long term approach should be to ingest the operation response in parts. This is beyond the scope of things we are doing but worth discussing with Sergey. The swagger schema defines what the response for a particular operation will be. This response can be very big but if we can split the response schema into multiple parts we can acquire them individually, summarize them and combine them all together. That would prevent us from having to spin up a new class to determine how to parse the json.

All of that being said, can we not achieve the same result by slimming down the actual swagger.json response for the operation?

Reply:
I don't want to modify the schema anymore than limiting it to APIs we want to leverage. Operations should come as defined from the publisher

Rest APIs are meant for publishers to communicate data in a standardized way, and once the person requesting that information gets the response, they can manipulate that data as they see fit for their application. In our case, leaving the properties in the GitHub schema indicates that these properties are still available if needed, but since we don't want them for our application, we cut them out on deserialization (kinda like model vs view state).

Long term, Sergey has work planned to parse and expose the response schema using the RestApiOperation object the kernel creates. What we do with that schema is still open for discussion (some ideas is using the AI to semantically identifying properties related to user intent and extracting only those fields from the response). Let's take this conversation offline and schedule a meeting to discuss all our ideas together.